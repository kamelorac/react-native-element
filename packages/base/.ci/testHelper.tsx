import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
