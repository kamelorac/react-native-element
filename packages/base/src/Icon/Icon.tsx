import React from 'react';
import {
  Platform,
  View,
  StyleSheet,
  ViewStyle,
  StyleProp,
  TextStyle,
  Pressable,
  ColorValue,
} from 'react-native';
import {
  IconButtonProps,
  IconProps as VectorIconProps,
} from 'react-native-vector-icons/Icon';
import Color from 'color';
import getIconType from '../helpers/getIconType';
import getIconStyle from '../helpers/getIconStyle';
import {
  androidRipple,
  defaultTheme,
  InlinePressableProps,
  RneFunctionComponent,
} from '../helpers';

export type IconType =
  | 'material'
  | 'material-community'
  | 'simple-line-icon'
  | 'zocial'
  | 'font-awesome'
  | 'octicon'
  | 'ionicon'
  | 'foundation'
  | 'evilicon'
  | 'entypo'
  | 'antdesign'
  | 'font-awesome-5'
  | string;

export interface IconObject {
  /** Name of icon. */
  name?: string;

  /** Color of icon. */
  color?: string;

  /** Size of icon. */
  size?: number;

  /** Type of icon */
  type?: IconType;

  /** Apply style to the icon using iconStyle. */
  iconStyle?: StyleProp<TextStyle>;
}

export type IconNode = boolean | React.ReactElement<{}> | Partial<IconProps>;

export interface IconProps extends InlinePressableProps, IconButtonProps {
  /** Type of icon set. [Supported sets here](#available-icon-sets). */
  type?: IconType;

  /** Update React Native Component.
   *  @default `Press handlers present then Pressable else View`
   */
  Component?: typeof React.Component;

  /** Reverses color scheme. */
  reverse?: boolean;

  /** Adds box shadow to button. */
  raised?: boolean;

  /** Add styling to container holding icon. */
  containerStyle?: StyleProp<ViewStyle>;

  /** Provide all props from react-native Icon component. */
  iconProps?: VectorIconProps;

  /** Specify reverse icon color. */
  reverseColor?: string;

  /** Disables onPress events. Only works when `onPress` has a handler. */
  disabled?: boolean;

  /** Style for the button when disabled. Only works when `onPress` has a handler. */
  disabledStyle?: StyleProp<ViewStyle>;

  /** Uses the solid font. */
  solid?: boolean;

  /** Uses the brands font (FontAwesome5 only). */
  brand?: boolean;
}

/** Icons are visual indicators usually used to describe action or intent.
 * They are also used for displaying information. */
export const Icon: RneFunctionComponent<IconProps> = ({
  type = 'material',
  name,
  size = 24,
  color: colorProp,
  iconStyle,
  iconProps,
  underlayColor = 'transparent',
  reverse = false,
  raised = false,
  containerStyle,
  reverseColor: reverseColorProp,
  disabled = false,
  disabledStyle,
  onPress,
  onLongPress,
  onPressIn,
  onPressOut,
  Component = onPress || onLongPress || onPressIn || onPressOut
    ? Pressable
    : View,
  solid = false,
  brand = false,
  theme = defaultTheme,
  pressableProps,
  ...rest
}) => {
  const color = colorProp || theme?.colors?.black;
  const reverseColor = reverseColorProp || theme?.colors?.white;
  const IconComponent = getIconType(type);
  const iconSpecificStyle = getIconStyle(type, { solid, brand });

  const getBackgroundColor = React.useMemo(() => {
    if (reverse) {
      return color;
    }
    return raised ? theme?.colors?.white : 'transparent';
  }, [color, raised, reverse, theme?.colors?.white]);

  const buttonStyles = React.useMemo(
    () => ({
      borderRadius: size + 4,
      height: size * 2 + 4,
      width: size * 2 + 4,
    }),
    [size]
  );

  return (
    <View
      style={StyleSheet.flatten([
        !raised && styles.container,
        (reverse || raised) && styles.button,
        (reverse || raised) && buttonStyles,
        raised && styles.raised,
        iconStyle && iconStyle.borderRadius
          ? {
              borderRadius: iconStyle.borderRadius,
            }
          : {},
        containerStyle && containerStyle,
      ])}
      testID="RNE__ICON__CONTAINER"
    >
      <Component
        testID="RNE__ICON__CONTAINER_ACTION"
        {...{
          android_ripple: androidRipple(
            Color(reverse ? color : (underlayColor as string))
              .alpha(0.3)
              .rgb()
              .string()
          ),
          onPress,
          onLongPress,
          onPressIn,
          onPressOut,
          disabled,
          accessibilityRole: 'button',
          ...pressableProps,
          ...rest,
        }}
      >
        <View
          style={StyleSheet.flatten([
            (reverse || raised) && buttonStyles,
            {
              backgroundColor: getBackgroundColor as ColorValue,
              alignItems: 'center',
              justifyContent: 'center',
            },
            disabled && styles.disabled,
            disabled && disabledStyle,
          ])}
          testID="RNE__ICON"
        >
          <IconComponent
            testID="RNE__ICON__Component"
            style={StyleSheet.flatten([
              { backgroundColor: 'transparent' },
              iconStyle && iconStyle,
            ])}
            size={size}
            name={name}
            color={reverse ? reverseColor : color}
            {...iconSpecificStyle}
            {...iconProps}
          />
        </View>
      </Component>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  button: {
    margin: 7,
  },
  raised: {
    ...Platform.select({
      android: {
        elevation: 2,
      },
      default: {
        shadowColor: 'rgba(0,0,0, .4)',
        shadowOffset: { height: 1, width: 1 },
        shadowOpacity: 1,
        shadowRadius: 1,
      },
    }),
  },
  disabled: {
    backgroundColor: '#D1D5D8',
  },
});

Icon.displayName = 'Icon';
