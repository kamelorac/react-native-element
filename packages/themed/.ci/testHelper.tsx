import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
