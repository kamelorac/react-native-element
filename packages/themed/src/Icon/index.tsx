import { withTheme } from '../config';
import {
  Icon,
  IconProps,
  IconNode,
  IconObject,
  IconType,
} from '@rneui/base/dist/Icon/Icon';

export { Icon };
export type { IconProps, IconNode, IconObject, IconType };
export default withTheme<IconProps>(Icon, 'Icon');
