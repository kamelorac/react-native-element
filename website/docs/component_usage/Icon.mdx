<!-- To display the available Icon set which is not commented in the component. -->

> Hint: use **reverse** to make your icon look like a button

## Available Icon Sets

The icon sets in React Native Elements are made possible through
[react-native-vector-icons](https://github.com/oblador/react-native-vector-icons).

The current list of available icons sets are:

- [antdesign](http://ant.design/components/icon/)
- [entypo](http://www.entypo.com/)
- [evilicon](http://evil-icons.io/)
- [feather](https://feathericons.com/)
- [font-awesome](https://fontawesome.com/v4.7.0/)
- [font-awesome-5](https://fontawesome.com/)
- [fontisto](https://www.fontisto.com/icons)
- [foundation](http://zurb.com/playground/foundation-icon-fonts-3)
- [ionicon](http://ionicons.com/)
- [material](https://material.io/tools/icons)
- [material-community](https://materialdesignicons.com/)
- [octicon](https://octicons.github.com/)
- [simple-line-icon](https://simplelineicons.github.io/)
- [zocial](http://weloveiconfonts.com/)

To check all the supported icons, visit [react-native-vector-icons directory](https://oblador.github.io/react-native-vector-icons/)

## Custom Icon Fonts

Register your own custom icons by calling
`registerCustomIconType('customid', customFont)`. Create a custom font by
following the
[ instructions for creating a custom font here](https://github.com/oblador/react-native-vector-icons#custom-fonts).
Also, you can use [Fontello](http://fontello.com/) to generate custom icon
fonts.

```SnackPlayer name=RNE Icon
import React from 'react';
import { View, Text } from 'react-native';
import { Icon } from '@rneui/themed';

export default () => {
return (
  <>
    <View
      style={{
        alignItems: 'center',
        paddingVertical: 5,
        flexGrow: 1,
      }}
    >
      <Icon
        name='rowing' />

      <Icon
        name='g-translate'
        color='#00aced' />

      <Icon
        name='sc-telegram'
        type='evilicon'
        color='#517fa4'
      />

      <Icon
        reverse
        name='ios-american-football'
        type='ionicon'
        color='#517fa4'
      />

      <Icon
        raised
        name='heartbeat'
        type='font-awesome'
        color='#f50'
        onPress={() => console.log('hello')} />
    </View>
  </>
);
};
```
