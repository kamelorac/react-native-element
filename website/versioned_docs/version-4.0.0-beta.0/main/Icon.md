---
id: icon

title: Icon

slug: /icon
---

import Usage from './usage/Icon/Icon.md'

Icons are visual indicators usually used to describe action or intent.

They are also used for displaying information.

## Usage

<Usage />

---

## Props

### Icon

- [Component](#component)
- [backgroundColor](#backgroundcolor)
- [borderRadius](#borderradius)
- [brand](#brand)
- [color](#color)
- [containerStyle](#containerstyle)
- [disabled](#disabled)
- [disabledStyle](#disabledstyle)
- [iconProps](#iconprops)
- [iconStyle](#iconstyle)
- [name](#name)
- [pressableProps](#pressableprops)
- [raised](#raised)
- [reverse](#reverse)
- [reverseColor](#reversecolor)
- [size](#size)
- [solid](#solid)
- [style](#style)
- [type](#type)

## Reference

### Icon

#### Component

Update React Native Component.

| Type            | Default |
| --------------- | ------- |
| React Component | None    |

---

#### backgroundColor

Background color of the button

| Type   | Default   |
| ------ | --------- |
| string | '#007AFF' |

---

#### borderRadius

Border radius of the button

Set to 0 to disable.

| Type   | Default |
| ------ | ------- |
| number | 5       |

---

#### brand

Uses the brands font (FontAwesome5 only).

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### color

Text and icon color

Use iconStyle or nest a Text component if you need different colors.

| Type   | Default |
| ------ | ------- |
| string | 'white' |

---

#### containerStyle

Add styling to container holding icon.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### disabled

If true, disable all interactions for this component.

Disables onPress events. Only works when `onPress` has a handler.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### disabledStyle

Style for the button when disabled. Only works when `onPress` has a handler.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### iconProps

Provide all props from react-native Icon component.

| Type      | Default |
| --------- | ------- |
| IconProps | None    |

---

#### iconStyle

Styles applied to the icon only

Good for setting margins or a different color.

| Type      | Default           |
| --------- | ----------------- |
| TextStyle | {marginRight: 10} |

---

#### name

Name of the icon to show

See [Icon Explorer app](https://github.com/oblador/react-native-vector-icons/tree/master/Examples/IconExplorer)


| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### pressableProps

| Type                                                                            | Default |
| ------------------------------------------------------------------------------- | ------- |
| Omit<PressableProps, "onPress" or "onLongPress" or "onPressIn" or "onPressOut"> | None    |

---

#### raised

Adds box shadow to button.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### reverse

Reverses color scheme.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### reverseColor

Specify reverse icon color.

| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### size

Size of the icon, can also be passed as fontSize in the style object.

| Type   | Default |
| ------ | ------- |
| number | 24      |

---

#### solid

Uses the solid font.

| Type    | Default |
| ------- | ------- |
| boolean | false   |

---

#### style

Style prop inherited from TextProps and TouchableWithoutFeedbackProperties

Only exist here so we can have ViewStyle or TextStyle

| Type                   | Default |
| ---------------------- | ------- |
| ViewStyle or TextStyle | None    |

---

#### type

Type of icon set. [Supported sets here](#available-icon-sets).

| Type   | Default  |
| ------ | -------- |
| string | material |

---
