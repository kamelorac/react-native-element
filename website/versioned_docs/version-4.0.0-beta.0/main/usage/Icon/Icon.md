import Snack from './snack/index.md'

<!-- To display the available Icon set which is not commented in the component. -->

> Hint: use **reverse** to make your icon look like a button

## Available Icon Sets

The icon sets in React Native Elements are made possible through
[react-native-vector-icons](https://github.com/oblador/react-native-vector-icons).

The current list of available icons sets are:

- [antdesign](http://ant.design/components/icon/)
- [entypo](http://www.entypo.com/)
- [evilicon](http://evil-icons.io/)
- [feather](https://feathericons.com/)
- [font-awesome](https://fontawesome.com/v4.7.0/)
- [font-awesome-5](https://fontawesome.com/)
- [fontisto](https://www.fontisto.com/icons)
- [foundation](http://zurb.com/playground/foundation-icon-fonts-3)
- [ionicon](http://ionicons.com/)
- [material](https://material.io/tools/icons)
- [material-community](https://materialdesignicons.com/)
- [octicon](https://octicons.github.com/)
- [simple-line-icon](https://simplelineicons.github.io/)
- [zocial](http://weloveiconfonts.com/)

To check all the supported icons, visit [react-native-vector-icons directory](https://oblador.github.io/react-native-vector-icons/)

## Custom Icon Fonts

Register your own custom icons by calling
`registerCustomIconType('customid', customFont)`. Create a custom font by
following the
[ instructions for creating a custom font here](https://github.com/oblador/react-native-vector-icons#custom-fonts).
Also, you can use [Fontello](http://fontello.com/) to generate custom icon
fonts.

<Snack />
