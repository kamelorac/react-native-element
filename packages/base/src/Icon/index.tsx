import { Icon, IconProps, IconNode, IconObject, IconType } from './Icon';

export { Icon };
export type { IconProps, IconNode, IconObject, IconType };
