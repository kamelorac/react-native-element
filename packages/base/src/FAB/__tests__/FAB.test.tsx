import React from 'react';
import { Animated, View } from 'react-native';
import { FAB } from '..';
import { renderWithWrapper, fireEvent } from '../../../.ci/testHelper';
import { Icon } from '../../Icon';

describe('FAB Component', () => {
  it('should match snapshot', () => {
    const { toJSON } = renderWithWrapper(<FAB title="Hey" size="small" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should have extended fab', () => {
    const { queryByText } = renderWithWrapper(<FAB title="test" />);
    expect(queryByText('test')).toBeTruthy();
  });

  it('should extended small fab', () => {
    const { wrapper, queryByText } = renderWithWrapper(
      <FAB title="test" size="small" />,
      'RNE_BUTTON_WRAPPER'
    );
    expect(queryByText('test')).toBeTruthy();
    expect(wrapper.findAllByType(View)[1].props.style).toMatchObject({
      height: 40,
    });
  });

  it('should have default size as large', () => {
    const { wrapper } = renderWithWrapper(<FAB />, 'RNE_BUTTON_WRAPPER');
    expect(wrapper.findAllByType(View)[1].props.style).toMatchObject({
      height: 56,
      width: 56,
    });
  });

  it('should render size small', () => {
    const { wrapper } = renderWithWrapper(
      <FAB size="small" />,
      'RNE_BUTTON_WRAPPER'
    );
    expect(wrapper.findAllByType(View)[1].props.style).toMatchObject({
      height: 40,
      width: 40,
    });
  });

  it('should render disabled fab', () => {
    const { wrapper } = renderWithWrapper(
      <FAB icon={{ name: 'edit' }} disabled={true} />
    );
    expect(wrapper.findByType(Icon).props).toMatchObject({
      name: 'edit',
    });
  });

  it('should handle onPress', () => {
    const onPress = jest.fn();
    const { queryByRole } = renderWithWrapper(<FAB onPress={onPress} />);
    const button = queryByRole('button');
    expect(button).toBeTruthy();
    fireEvent.press(button!);
    expect(onPress).toBeCalledTimes(1);
  });

  it('should use placement left', () => {
    const { wrapper } = renderWithWrapper(<FAB placement="left" />);
    expect(wrapper.findByType(Animated.View).props.style).toMatchObject({
      position: 'absolute',
      bottom: 0,
      left: 0,
    });
  });

  it('should use placement right', () => {
    const { wrapper } = renderWithWrapper(<FAB placement="right" />);
    expect(wrapper.findByType(Animated.View).props.style).toMatchObject({
      position: 'absolute',
      bottom: 0,
      right: 0,
    });
  });
});
