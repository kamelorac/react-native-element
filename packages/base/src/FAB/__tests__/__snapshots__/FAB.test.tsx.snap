// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAB Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderRadius": 28,
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          {
            "borderRadius": 28,
            "elevation": 4,
          },
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#ad1457",
              "borderColor": "#2089dc",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "#ffffff",
                "fontSize": 18,
                "marginHorizontal": 8,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Hey
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
